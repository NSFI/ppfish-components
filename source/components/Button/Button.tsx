import * as React from 'react';
import { findDOMNode } from 'react-dom';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../Icon';
import Group from './ButtonGroup';

const rxTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;
const isTwoCNChar = rxTwoCNChar.test.bind(rxTwoCNChar);
function isString(str: any) {
  return typeof str === 'string';
}

// Insert one space between two chinese characters automatically.
function insertSpace(child: React.ReactChild, needInserted: boolean) {
  // Check the child if is undefined or null.
  if (child == null) {
    return;
  }
  const SPACE = needInserted ? ' ' : '';
  // strictNullChecks oops.
  if (typeof child !== 'string' && typeof child !== 'number' &&
    isString(child.type) && isTwoCNChar(child.props.children)) {
    return React.cloneElement(child, {},
      child.props.children.split('').join(SPACE));
  }
  if (typeof child === 'string') {
    if (isTwoCNChar(child)) {
      child = child.split('').join(SPACE);
    }
    return <span>{child}</span>;
  }
  return child;
}

export type ButtonType = 'default' | 'primary' | 'ghost' | 'dashed' | 'danger';
export type ButtonShape = 'circle' | 'circle-outline';
export type ButtonSize = 'small' | 'default' | 'large';
export type ButtonHTMLType = 'submit' | 'button' | 'reset';

export interface BaseButtonProps {
  type?: ButtonType;
  icon?: string;
  shape?: ButtonShape;
  size?: ButtonSize;
  loading?: boolean | { delay?: number };
  prefixCls?: string;
  className?: string;
  ghost?: boolean;
}

export type AnchorButtonProps = {
  href: string;
  target?: string;
  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
} & BaseButtonProps & React.AnchorHTMLAttributes<HTMLAnchorElement>;

export type NativeButtonProps = {
  htmlType?: ButtonHTMLType;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
} & BaseButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = AnchorButtonProps | NativeButtonProps;

export default class Button extends React.Component<ButtonProps, any> {
  static Group: typeof Group;
  static __ANT_BUTTON = true;

  static defaultProps = {
    prefixCls: 'ant-btn',
    loading: false,
    ghost: false,
  };

  static propTypes = {
    type: PropTypes.string,
    shape: PropTypes.oneOf(['circle', 'circle-outline']),
    size: PropTypes.oneOf(['large', 'default', 'small']),
    htmlType: PropTypes.oneOf(['submit', 'button', 'reset']),
    onClick: PropTypes.func,
    loading: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    className: PropTypes.string,
    icon: PropTypes.string,
  };

  timeout: number;
  delayTimeout: number;

  constructor(props: ButtonProps) {
    super(props);
    this.state = {
      loading: props.loading,
      clicked: false,
      hasTwoCNChar: false,
    };
  }

  componentDidMount() {
    this.fixTwoCNChar();
  }

  componentWillReceiveProps(nextProps: ButtonProps) {
    const currentLoading = this.props.loading;
    const loading = nextProps.loading;

    if (currentLoading) {
      clearTimeout(this.delayTimeout);
    }

    if (typeof loading !== 'boolean' && loading && loading.delay) {
      this.delayTimeout = window.setTimeout(() => this.setState({ loading }), loading.delay);
    } else {
      this.setState({ loading });
    }
  }

  componentDidUpdate() {
    this.fixTwoCNChar();
  }

  componentWillUnmount() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
    if (this.delayTimeout) {
      clearTimeout(this.delayTimeout);
    }
  }

  fixTwoCNChar() {
    // Fix for HOC usage like <FormatMessage />
    const node = (findDOMNode(this) as HTMLElement);
    const buttonText = node.textContent || node.innerText;
    if (this.isNeedInserted() && isTwoCNChar(buttonText)) {
      if (!this.state.hasTwoCNChar) {
        this.setState({
          hasTwoCNChar: true,
        });
      }
    } else if (this.state.hasTwoCNChar) {
      this.setState({
        hasTwoCNChar: false,
      });
    }
  }

  handleClick: React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement> = e => {
    // Add click effect
    this.setState({ clicked: true });
    clearTimeout(this.timeout);
    this.timeout = window.setTimeout(() => this.setState({ clicked: false }), 500);

    const onClick = this.props.onClick;
    if (onClick) {
      (onClick as React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)(e);
    }
  }

  isNeedInserted() {
    const { icon, children } = this.props;
    return React.Children.count(children) === 1 && !icon;
  }

  render() {
    const {
      type, shape, size, className, children, icon, prefixCls, ghost, loading: _loadingProp, ...rest
    } = this.props;

    const { loading, clicked, hasTwoCNChar } = this.state;

    // large => lg
    // small => sm
    let sizeCls = '';
    switch (size) {
      case 'large':
        sizeCls = 'lg';
        break;
      case 'small':
        sizeCls = 'sm';
      default:
        break;
    }

    const classes = classNames(prefixCls, className, {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${shape}`]: shape,
      [`${prefixCls}-${sizeCls}`]: sizeCls,
      [`${prefixCls}-icon-only`]: !children && icon,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-clicked`]: clicked,
      [`${prefixCls}-background-ghost`]: ghost,
      [`${prefixCls}-two-chinese-chars`]: hasTwoCNChar,
    });

    const iconType = loading ? 'loading' : icon;
    const iconNode = iconType ? <Icon type={iconType} /> : null;
    const kids = (children || children === 0)
      ? React.Children.map(children, child => insertSpace(child, this.isNeedInserted())) : null;

    if ('href' in rest) {
      return (
        <a
          {...rest}
          className={classes}
          onClick={this.handleClick}
        >
        {iconNode}{kids}
        </a>
      );
    } else {
      // React does not recognize the `htmlType` prop on a DOM element. Here we pick it out of `rest`.
      const { htmlType, ...otherProps } = rest;

      return (
        <button
          {...otherProps}
          type={htmlType || 'button'}
          className={classes}
          onClick={this.handleClick}
        >
        {iconNode}{kids}
        </button>
      );
    }
  }
}
