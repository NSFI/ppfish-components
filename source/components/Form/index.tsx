import Form from './Form';
import './style/index.less';

export {
  FormProps,
  FormComponentProps,
  FormCreateOption,
  ValidateCallback,
  ValidationRule
} from './Form';
export { FormListProps } from './FormList';
export { FormItemProps } from './FormItem';

export default Form;
