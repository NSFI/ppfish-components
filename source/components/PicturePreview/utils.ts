import React from 'react';

export type ImageItem = {
  rotateValue?: number;
  src?: string;
  alt?: string;
  name?: string;
};

let conMaxWidth = 1024;
let conMaxHeight = 768;
const CON_MAX_WIDTH = conMaxWidth > window.innerWidth ? window.innerWidth : conMaxWidth; //容器最大宽度
const CON_MIN_WIDTH = 360; //容器最小宽度
const CON_MAX_HEIGHT = conMaxHeight > window.innerHeight ? window.innerHeight : conMaxHeight; //容器最大高度
const CON_MIN_HEIGHT = 360; //容器最小高度
const DEFAULT_RATIO = 0.8; //默认的图片显示比例

/**
 * 工具函数  start
 */
export function num2px(num) {
  return parseInt(num, 10) + 'px';
}

export function px2num(str) {
  return Number(str.replace('px', '')) || 0;
}

/**
 * el1元素的区域是否超过el2元素
 * @param  {[type]}  el1 [description]
 * @param  {[type]}  el2 [description]
 * @return {Boolean}     [description]
 */
export function isLarger(el1, el2) {
  return el1.clientHeight > el2.clientHeight || el1.clientWidth > el2.clientWidth;
}

export function isSourceChange(oldSource: ImageItem[], newSource: ImageItem[]): boolean {
  return JSON.stringify(oldSource) !== JSON.stringify(newSource);
}

export const setStyle = (el, css) => {
  for (let key in css) {
    el.style[key] = css[key];
  }
};

export const getImageSize = function (
  image: ImageItem & HTMLImageElement,
  callback: (naturalWidth: number, naturalHeight: number) => void,
  scope?: any,
) {
  let newImage;
  if (!image.src) {
    callback.call(scope, 0, 0);
  } else if (image.naturalWidth) {
    // 现代浏览器
    callback.call(scope, image.naturalWidth, image.naturalHeight);
  } else {
    // 低版本浏览器
    newImage = document.createElement('img');
    newImage.onload = function () {
      callback.call(scope, this.width, this.height);
    };
    newImage.src = image.src;
  }
};

export const isOne2One = image => {
  return Math.round(image.ratio * 100) === 100;
};

let downloadImgUrl: string | null = null;

export const downloadFn = (url, alt) => {
  let getBlobImage = img => {
    let canvas = document.createElement('canvas'),
      ctx = canvas.getContext('2d');
    canvas.width = img.width;
    canvas.height = img.height;
    ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
    return new Promise<Blob>(resolve => {
      canvas.toBlob(blob => {
        resolve(blob);
      });
    });
  };

  let img = new Image();
  img.crossOrigin = 'anonymous';
  img.onload = () => {
    let promise = getBlobImage(img);
    promise.then(blob => {
      let dLink = document.createElement('a');
      dLink.download = alt || '';
      downloadImgUrl = window.URL.createObjectURL(blob);
      dLink.href = downloadImgUrl;
      dLink.onclick = () => {
        window.requestAnimationFrame(() => {
          window.URL.revokeObjectURL(downloadImgUrl);
          downloadImgUrl = null;
        });
      };
      dLink.click();
    });
  };
  // 在URL后添加随机数以避免浏览器缓存，使crossOrigin生效
  img.src = url;
};

export const isFullEnabled = () => {
  return (
    document.fullscreenEnabled ||
    document['mozFullScreenEnabled'] ||
    document['webkitFullscreenEnabled'] ||
    document['msFullscreenEnabled']
  );
};

export const calculateImg = (naturalWidth, naturalHeight) => {
  let width, height, imgRatio;

  if (naturalWidth == 0 || naturalHeight == 0) {
    width = CON_MIN_WIDTH;
    height = CON_MIN_HEIGHT;
    imgRatio = 0;
  } else {
    //计算容器的宽度
    width = naturalWidth * DEFAULT_RATIO; //默认0.8倍显示图片
    if (width > CON_MAX_WIDTH) {
      width = CON_MAX_WIDTH;
    } else if (width < CON_MIN_WIDTH) {
      width = CON_MIN_WIDTH;
    }

    //计算图片的缩放比例
    imgRatio = (naturalWidth && width / naturalWidth) || 0;

    //计算容器的高度
    height = naturalHeight * imgRatio;
    if (height > CON_MAX_HEIGHT) {
      height = CON_MAX_HEIGHT;
    } else if (height < CON_MIN_HEIGHT) {
      height = CON_MIN_HEIGHT;
    }
  }
  return { width, height, imgRatio };
};

export const getImgByChildren = (children): { name: string; src: string }[] | undefined => {
  if (children) {
    return React.Children.map(children, (child: React.ReactElement<HTMLImageElement>) => {
      let img = {
        name: '',
        src: '',
      };

      if (child.type === 'img') {
        img.name = child.props.name || child.props.alt;
        img.src = child.props.src;
      }

      return img;
    }).filter(item => item);
  }
};

export const getImgByProps = (source, children, imgs) => {
  let _imgs = null;
  if (source && source.length) {
    let sourceStr = JSON.stringify(source);
    if (sourceStr !== JSON.stringify(imgs)) {
      _imgs = JSON.parse(sourceStr);
    }
  } else if (children) {
    _imgs = getImgByChildren(children);
  }
  return _imgs;
};
