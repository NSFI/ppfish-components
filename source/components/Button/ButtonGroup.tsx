import * as React from 'react';
import classNames from 'classnames';
import { ButtonSize } from './Button';

export interface ButtonGroupProps {
  size?: ButtonSize;
  style?: React.CSSProperties;
  className?: string;
  prefixCls?: string;
}

const ButtonGroup: React.SFC<ButtonGroupProps> = (props) => {
  const { prefixCls = 'fishd-btn-group', size, className, ...others } = props;

  // large => lg
  // small => sm
  let sizeCls = '';
  switch (size) {
    case 'large':
      sizeCls = 'lg';
      break;
    case 'small':
      sizeCls = 'sm';
    default:
      break;
  }

  const classes = classNames(prefixCls, {
    [`${prefixCls}-${sizeCls}`]: sizeCls,
  }, className);

  return <div {...others} className={classes} />;
};

export default ButtonGroup;
