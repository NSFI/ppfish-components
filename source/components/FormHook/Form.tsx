import  React, { useMemo } from 'react';
import classNames from 'classnames';
import FieldForm, { List } from './rcFieldForm';
import { FormProps as RcFormProps } from './rcFieldForm/Form';
import { ValidateErrorEntity } from './rcFieldForm/interface';
import { Options } from 'scroll-into-view-if-needed';
import { ColProps } from '../Grid';
import { FormContext, FormContextProps } from './context';
import { FormLabelAlign } from './interface';
import useForm, { FormInstance } from './hooks/useForm';
export type RequiredMark = boolean | 'optional';
export type FormLayout = 'horizontal' | 'inline' | 'vertical';

/**
 * 没有添加 Size  需要有一个全局的 SizeContext, 所有组件都要配合
 * config context 中的全局配置未添加 getPrefixCls, direction, form 等情况
 * RequiredMark - optional, rtl 暂不实现
 * colon 特意删除
 *
 * antd 中 form 的表单标签可换行功能 需要 Col 组件的支持
 *
 * 缺少 Space 组件
 *
 * 因为 Col 组件的不一样, 导致 form 的 layout 即使默认是horizontal, 如果没传span, 也会和 vertical 一样
 *
 * FormItem hasFeedback 属性没有 FormItemStatusContext 支持, 这里使用老的一套 有.has-feedback属性 input 则显示
 *
 * 置空存在在现在的Col 组件上存在动画问题, 查看高级搜索 demo, 去除 fields 样式
 *
 * DatePicker 中不能选择月, 待加强
 *
 * DateRangePicker 第一次点击时有 warning 需要处理  `DateTable: prop type `rangeState.endDate` is invalid`
 */

export interface FormProps<Values = any> extends Omit<RcFormProps<Values>, 'form'> {
  prefixCls?: string;
  name?: string;
  layout?: FormLayout;
  labelAlign?: FormLabelAlign;
  labelWrap?: boolean;
  labelCol?: ColProps;
  wrapperCol?: ColProps;
  form?: FormInstance<Values>;
  scrollToFirstError?: Options | boolean;
  hideRequiredMark?: boolean;
}

const InternalForm: React.ForwardRefRenderFunction<FormInstance, FormProps> = (props, ref) => {

  const {
    prefixCls,
    className = '',
    form,
    labelAlign,
    labelWrap,
    labelCol,
    wrapperCol,
    hideRequiredMark,
    layout = 'horizontal',
    scrollToFirstError,
    onFinishFailed,
    name,
    ...restFormProps
  } = props;

  const formClassName = classNames(
    prefixCls,
    {
      [`${prefixCls}-horizontal`]: layout === 'horizontal',
      [`${prefixCls}-vertical`]: layout === 'vertical',
      [`${prefixCls}-inline`]: layout === 'inline',
      [`${prefixCls}-hide-required-mark`]: hideRequiredMark,
    },
    className,
  );

  const [wrapForm] = useForm(form);
  const { __INTERNAL__ } = wrapForm;
  __INTERNAL__.name = name;

  const formContextValue = useMemo<FormContextProps>(
    () => ({
      name,
      labelAlign,
      labelCol,
      labelWrap,
      wrapperCol,
      vertical: layout === 'vertical',
      requiredMark: hideRequiredMark,
      itemRef: __INTERNAL__.itemRef,
    }),
    [name, labelAlign, labelCol, wrapperCol, layout, hideRequiredMark],
  );

  React.useImperativeHandle(ref, () => wrapForm);

  const onInternalFinishFailed = (errorInfo: ValidateErrorEntity) => {
    onFinishFailed?.(errorInfo);

    let defaultScrollToFirstError: Options = { block: 'nearest' };

    if (scrollToFirstError && errorInfo.errorFields.length) {
      if (typeof scrollToFirstError === 'object') {
        defaultScrollToFirstError = scrollToFirstError;
      }
      wrapForm.scrollToField(errorInfo.errorFields[0].name, defaultScrollToFirstError);
    }
  };

  return (
      <FormContext.Provider value={formContextValue}>
        <FieldForm
          id={name}
          {...restFormProps}
          name={name}
          onFinishFailed={onInternalFinishFailed}
          form={wrapForm}
          className={formClassName}
        />
      </FormContext.Provider>
  );
};




const Form = React.forwardRef<FormInstance, FormProps>(InternalForm) as <Values = any>(
  props: React.PropsWithChildren<FormProps<Values>> & { ref?: React.Ref<FormInstance<Values>> },
) => React.ReactElement;

(Form as any).defaultProps = {
  prefixCls: 'fishd-formHook',
  layout: 'horizontal',
  labelAlign: 'left', // 按照 Form 的标准 默认靠左
  hideRequiredMark: false,
  onSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
  },
}


export { useForm, List, FormInstance };

export default Form;
