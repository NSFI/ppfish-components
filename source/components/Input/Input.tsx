import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import { Omit } from '../../utils/type';

function fixControlledValue<T>(value: T) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export interface InputProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'> {
  prefixCls?: string;
  size?: 'large' | 'default' | 'small';
  onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
  addonBefore?: React.ReactNode;
  addonAfter?: React.ReactNode;
  prefix?: React.ReactNode;
  suffix?: React.ReactNode;
}

export default class Input extends React.Component<InputProps, any> {
  static Group: typeof Group;
  static Search: typeof Search;
  static TextArea: typeof TextArea;

  static defaultProps = {
    prefixCls: 'fishd-input',
    type: 'text',
    disabled: false,
  };

  static propTypes = {
    type: PropTypes.string,
    id: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    size: PropTypes.oneOf(['small', 'default', 'large']),
    maxLength: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
    disabled: PropTypes.bool,
    value: PropTypes.any,
    defaultValue: PropTypes.any,
    className: PropTypes.string,
    addonBefore: PropTypes.node,
    addonAfter: PropTypes.node,
    prefixCls: PropTypes.string,
    autosize: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    onPressEnter: PropTypes.func,
    onKeyDown: PropTypes.func,
    onKeyUp: PropTypes.func,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    prefix: PropTypes.node,
    suffix: PropTypes.node,
  };

  input: HTMLInputElement;

  handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { onPressEnter, onKeyDown } = this.props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  }

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  getInputClassName() {
    const { prefixCls, size, disabled } = this.props;
    return classNames(prefixCls, {
      [`${prefixCls}-sm`]: size === 'small',
      [`${prefixCls}-lg`]: size === 'large',
      [`${prefixCls}-disabled`]: disabled,
    });
  }

  saveInput = (node: HTMLInputElement) => {
    this.input = node;
  }

  renderLabeledInput(children: React.ReactElement<any>) {
    const props = this.props;
    // Not wrap when there is not addons
    if ((!props.addonBefore && !props.addonAfter)) {
      return children;
    }

    const wrapperClassName = `${props.prefixCls}-group`;
    const addonClassName = `${wrapperClassName}-addon`;
    const addonBefore = props.addonBefore ? (
      <span className={addonClassName}>
        {props.addonBefore}
      </span>
    ) : null;

    const addonAfter = props.addonAfter ? (
      <span className={addonClassName}>
        {props.addonAfter}
      </span>
    ) : null;

    const className = classNames(`${props.prefixCls}-wrapper`, {
      [wrapperClassName]: (addonBefore || addonAfter),
    });

    const groupClassName = classNames(`${props.prefixCls}-group-wrapper`, {
      [`${props.prefixCls}-group-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-group-wrapper-lg`]: props.size === 'large',
    });

    // Need another wrapper for changing display:table to display:inline-block
    // and put style prop in wrapper
    if (addonBefore || addonAfter) {
      return (
        <span
          className={groupClassName}
          style={props.style}
        >
          <span className={className}>
            {addonBefore}
            {React.cloneElement(children, { style: null })}
            {addonAfter}
          </span>
        </span>
      );
    }
    return (
      <span className={className}>
        {addonBefore}
        {children}
        {addonAfter}
      </span>
    );
  }

  renderLabeledIcon(children: React.ReactElement<any>) {
    const { props } = this;
    if (!('prefix' in props || 'suffix' in props)) {
      return children;
    }

    const prefix = props.prefix ? (
      <span className={`${props.prefixCls}-prefix`}>
        {props.prefix}
      </span>
    ) : null;

    const suffix = props.suffix ? (
      <span className={`${props.prefixCls}-suffix`}>
        {props.suffix}
      </span>
    ) : null;

    const affixWrapperCls = classNames(props.className, `${props.prefixCls}-affix-wrapper`, {
      [`${props.prefixCls}-affix-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-affix-wrapper-lg`]: props.size === 'large',
    });
    return (
      <span
        className={affixWrapperCls}
        style={props.style}
      >
        {prefix}
        {React.cloneElement(children, { style: null, className: this.getInputClassName() })}
        {suffix}
      </span>
    );
  }

  renderInput() {
    const { value, className } = this.props;
    // Fix https://fb.me/react-unknown-prop
    const otherProps = omit(this.props, [
      'prefixCls',
      'onPressEnter',
      'addonBefore',
      'addonAfter',
      'prefix',
      'suffix',
    ]);

    if ('value' in this.props) {
      otherProps.value = fixControlledValue(value);
      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      delete otherProps.defaultValue;
    }
    return this.renderLabeledIcon(
      <input
        {...otherProps}
        className={classNames(this.getInputClassName(), className)}
        onKeyDown={this.handleKeyDown}
        ref={this.saveInput}
      />,
    );
  }

  render() {
    return this.renderLabeledInput(this.renderInput());
  }
}
