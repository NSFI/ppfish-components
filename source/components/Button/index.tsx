import Button from './Button';
import ButtonGroup from './ButtonGroup';

export { ButtonProps, ButtonShape, ButtonSize, ButtonType } from './Button';
export { ButtonGroupProps } from './ButtonGroup';

Button.Group = ButtonGroup;
export default Button;
