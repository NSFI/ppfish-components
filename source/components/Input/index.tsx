import Input from './Input';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import './style/index.less';

export {InputProps} from './Input';
export {GroupProps} from './Group';
export {SearchProps} from './Search';
export {TextAreaProps} from './TextArea';

Input.Group = Group;
Input.Search = Search;
Input.TextArea = TextArea;
export default Input;
