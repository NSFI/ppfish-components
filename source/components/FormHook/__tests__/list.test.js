import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Form from '..';
import Input from '../../Input';
import Button from '../../Button';
import { sleep } from '../../../../tools/tests/utils';

describe('Form.List', () => {
  async function change(wrapper, index, value) {
    wrapper.find(Input).at(index).simulate('change', { target: { value } });
    await sleep();
    wrapper.update();
  }

  function testList(name, renderField) {
    it(name, async () => {
      jest.useFakeTimers();

      const wrapper = mount(
        <Form>
          <Form.List name="list">
            {(fields, { add, remove }) => (
              <>
                {fields.map(field => renderField(field))}
                <Button className="add" onClick={add}>
                  Add
                </Button>
                <Button
                  className="remove-0"
                  onClick={() => {
                    remove(0);
                  }}
                />
                <Button
                  className="remove-1"
                  onClick={() => {
                    remove(1);
                  }}
                />
              </>
            )}
          </Form.List>
        </Form>,
      );

      function operate(className) {
        act(() => {
          wrapper.find(className).last().simulate('click');
          jest.runAllTimers();
        });
        wrapper.update();
      }

      operate('.add');
      expect(wrapper.find(Input).length).toBe(1);

      operate('.add');
      expect(wrapper.find(Input).length).toBe(2);

      operate('.add');
      expect(wrapper.find(Input).length).toBe(3);

      await change(wrapper, 2, '');
      act(() => {
        jest.runAllTimers();
      });
      wrapper.update();
      expect(wrapper.find('.fishd-formHook-item-explain div').length).toBe(1);

      operate('.remove-0');
      expect(wrapper.find(Input).length).toBe(2);
      expect(wrapper.find('.fishd-formHook-item-explain div').length).toBe(1);

      operate('.remove-1');
      expect(wrapper.find(Input).length).toBe(1);
      expect(wrapper.find('.fishd-formHook-item-explain div').length).toBe(0);

      jest.useRealTimers();
    });
  }

  testList('operation correctly', field => (
    <Form.Item {...field} rules={[{ required: true }]}>
      <Input />
    </Form.Item>
  ));

  testList('nest noStyle', field => (
    <Form.Item key={field.key}>
      <Form.Item noStyle {...field} rules={[{ required: true }]}>
        <Input />
      </Form.Item>
    </Form.Item>
  ));

  it('correct onFinish values', async () => {
    async function click(wrapper, className) {
      wrapper.find(className).last().simulate('click');
      await sleep();
      wrapper.update();
    }

    const onFinish = jest.fn().mockImplementation(() => {});

    const wrapper = mount(
      <Form
        onFinish={v => {
          if (typeof v.list[0] === 'object') {
            /* old version led to SyntheticEvent be passed as an value here
              that led to weird infinite loop somewhere and OutOfMemory crash */
            v = new Error('We expect value to be a primitive here');
          }
          onFinish(v);
        }}
      >
        <Form.List name="list">
          {(fields, { add, remove }) => (
            <>
              {fields.map(field => (
                // key is in a field
                // eslint-disable-next-line react/jsx-key
                <Form.Item {...field}>
                  <Input />
                </Form.Item>
              ))}
              <Button className="add" onClick={add}>
                Add
              </Button>
              <Button className="remove" onClick={() => remove(0)}>
                Remove
              </Button>
            </>
          )}
        </Form.List>
      </Form>,
    );

    await click(wrapper, '.add');
    await change(wrapper, 0, 'input1');
    wrapper.find('form').simulate('submit');
    await sleep();
    expect(onFinish).toHaveBeenLastCalledWith({ list: ['input1'] });

    await click(wrapper, '.add');
    await change(wrapper, 1, 'input2');
    await click(wrapper, '.add');
    await change(wrapper, 2, 'input3');
    wrapper.find('form').simulate('submit');
    await sleep();
    expect(onFinish).toHaveBeenLastCalledWith({ list: ['input1', 'input2', 'input3'] });

    await click(wrapper, '.remove'); // will remove first input
    wrapper.find('form').simulate('submit');
    await sleep();
    expect(onFinish).toHaveBeenLastCalledWith({ list: ['input2', 'input3'] });
  });

  it('list errors', async () => {
    jest.useFakeTimers();

    let operation;
    const wrapper = mount(
      <Form>
        <Form.List
          name="list"
          rules={[
            {
              validator: async (_, value) => {
                if (value.length < 2) {
                  return Promise.reject(new Error('At least 2'));
                }
              },
            },
          ]}
        >
          {(_, opt, { errors }) => {
            operation = opt;
            return <Form.ErrorList errors={errors} />;
          }}
        </Form.List>
      </Form>,
    );

    async function addItem() {
      await act(async () => {
        operation.add();
        await sleep(100);
        jest.runAllTimers();
        wrapper.update();
      });
    }

    await addItem();
    expect(wrapper.find('.fishd-formHook-item-explain div').text()).toEqual('At least 2');

    await addItem();
    expect(wrapper.find('.fishd-formHook-item-explain div')).toHaveLength(0);

    jest.useRealTimers();
  });

  it('should render empty without errors', () => {
    const wrapper = mount(<Form.ErrorList />);
    expect(wrapper.render()).toMatchSnapshot();
  });
});
