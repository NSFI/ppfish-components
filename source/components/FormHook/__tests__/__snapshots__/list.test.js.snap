// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.List should render empty without errors 1`] = `null`;
